ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) struct FuncIndex(u16);
+
+#[derive(Copy, Clone, Debug, Default, PartialEq)]
+pub(crate) struct ExtraArgsIndex(u16);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub(crate) struct TypeIndex(u16);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) struct ConstIndex(u16);
+
+impl ConstIndex {
+    pub(crate) fn get(&self) -> u16 {
+        self.0
+    }
 }
 
-impl Operand {
-    pub(crate) fn new(kind: OpKind, val: u64) -> Self {
-        // check if the operand's value can fit in a short operand.
-        if val <= SHORT_OPERAND_MAX_VALUE {
-            Self::Short(ShortOperand::new(kind, val))
-        } else {
-            todo!()
+/// The packed representation of an instruction operand.
+///
+/// # Encoding
+///
+/// ```ignore
+///  1             15
+/// +---+--------------------------+
+/// | k |         index            |
+/// +---+--------------------------+
+/// ```
+///
+///  - `k=0`: `index` is a local variable index
+///  - `k=1`: `index` is a constant index
+///
+///  The IR can represent 2^{15} = 32768 locals, and as many constants.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub(crate) struct PackedOperand(u16);
+
+impl PackedOperand {
+    pub fn new(op: &Operand) -> Self {
+        match op {
+            Operand::Local(lidx) => {
+                debug_assert!(lidx.get() <= MAX_OPERAND_IDX);
+                PackedOperand(lidx.get())
+            }
+            Operand::Const(constidx) => {
+                debug_assert!(constidx.get() <= MAX_OPERAND_IDX);
+                PackedOperand(constidx.get() | !OPERAND_INDEX_MASK)
+            }
         }
     }
 
-    fn raw(&self) -> u64 {
-        match self {
-            Self::Long(_) => todo!(),
-            Self::Short(op) => op.0,
+    /// Unpacks a [PackedOperand] into a [Operand].
+    pub fn get(&self) -> Operand {
+        if (self.0 & !OPERAND_INDEX_MASK) == 0 {
+            Operand::Local(InstrIndex(self.0))
+        } else {
+            Operand::Const(ConstIndex(self.0 & OPERAND_INDEX_MASK))
         }
     }
+}
 
-    fn kind(&self) -> OpKind {
+/// An unpacked representation of a operand.
+///
+/// This exists both as a convenience (as working with packed operands is laborious) and as a means
+/// to add type safety when using operands.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum Operand {
+    Local(InstrIndex),
+    Const(ConstIndex),
+}
+
+impl fmt::Display for Operand {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            Self::Long(_) => todo!(),
-            Self::Short(op) => op.kind(),
+            Self::Local(idx) => write!(f, "%{}", idx.get()),
+            Self::Const(idx) => write!(f, "Const({})", idx.get()), // FIXME
         }
     }
+}
+
+// FIXME: this isn't the correct representation of a constant.
+// It should be a bag of bytes and a type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum Constant {
+    Usize(usize),
```

Comment:
Note that this `enum` is now two machine words big. That's not a problem for now, but given that most usize` constants will be tiny (my guess is that > 50% will be the value `0`!) wasteful. Also some constants won't have a fixed size (e.g. `String)`. So at some point we'll want something more compact for the common case. I'm not sure if that's what's meant by "bag of bytes and a type" or not.

Reply:
Agreed that this could be more compact, but I think this part of the IR should be easy to change without breaking lots of other parts of the IR.

> I'm not sure if that's what's meant by "bag of bytes and a type" or not.

A design I've seen in the past for representing constants is:
```
struct Const {
  ty: Type,
  bytes: Vec<u8>,
}
```

And the type tells you how to interpret the bytes.

A more optimal version would be:
```
constant_data: Vec<u8>; // somewhere in the module

struct Const {
  ty: Type,
  bytes: &[u8]
}
```

Or:
```
```
constant_data: Vec<u8>; // somewhere in the module

struct Const {
  ty: Type,
  cdata_idx: usize,
  cdata_len: usize, // maybe not needed, but you get the idea.
}
```

Reply:
I think we can resolve this for now, and think more about it later. Because most constants are small (`0` is almost certainly the most common constant!), though, I would prefer they don't all have to heap allocate.